Support multiple firmware protocols

aedsinger opened the pull request:
This PR allows the Wacc, Pimu, and Stepper classes to handle multiple protocols. This works by

* Reading the current protocol off the uC firmware
* Changing the base clase of the device (eg Pimu) from PimuBase to Pimu_Protocol_P0

This decomposition allows new firmware functionality to coexist at the Stretch Body level with older firmware functionality.

This PR has been tested against the firmware branch:

https://github.com/hello-robot/stretch_firmware/tree/bugfix/remove_sync_timestamps

It currently supports the P0 and P1, although for P1 the waypoint trajectory functionality hasn't bee surfaced in stepper.py

aedsinger commented:
Still todo:
* Unit tests
* Update RE1_firmware_updater.py so that it provides multiple protocol support
* Graceful error reporting if a user calls a, for example, P1 method, on a P0 instance

hello-binit commented:
@aedsinger, I think this branch has conflicts with feature/firmware_protocol because I rebased it with develop yesterday. You might want to start use interactive rebase to drop the 6 conflicting commits (135b76f^..6ade454).

aedsinger commented:
I'm still learning rebasing and couldn't quite figure out how to make it work. In the end I merged....

aedsinger commented:
Still todo:

    Unit tests
    Graceful error reporting if a user calls a, for example, P1 method, on a P0 instance

hello-binit commented on body/stretch_body/pimu.py:
```diff
@@ -426,30 +375,32 @@ def pack_trigger(self,s,sidx):
             return sidx
 
+    def unpack_status(self,s):
+        pass
```
Would be better to `raise NotImplementedError()` here.

hello-binit commented on body/stretch_body/pimu.py:
```diff
@@ -488,3 +437,113 @@ def step_sentry(self):
+                Disabling device.
+                Please upgrade the firmware and/or version of Stretch Body.
+                ----------------
+                """.format(self.name, self.board_info['protocol_version'], self.supported_protocols)
```
Should be `self.supported_protocols.keys()`

aedsinger replied:
Fixed

hello-binit commented on body/stretch_body/stepper.py:
```diff
@@ -680,8 +632,11 @@ def pack_trigger(self, s, sidx):
             return sidx
 
+    def unpack_status(self,s):
+        pass
```
Would be better to `raise NotImplementedError()` here.

aedsinger replied:
Fixed

hello-binit commented on body/stretch_body/pimu.py:
```diff
@@ -351,41 +335,6 @@ def unpack_board_info(self,s):
-            self.status['cpu_temp']=self.get_cpu_temp()
-            return sidx
-
     def pack_config(self,s,sidx):
```
`pack_config` differs in P0 and P1: P1 introduces the "enable_sync_mode" config to PIMU. `pack_config` should move to the Protocol_Px classes.

aedsinger replied:
The 'new' P1 under bugfix/remove_sync_timestamps doesn't include enable_sync_mode so leaving pack_config out of Px for now.

hello-binit commented on body/stretch_body/wacc.py:
```diff
@@ -215,30 +179,110 @@ def pack_config(self,s,sidx):
             sidx += 1
```
`pack_config` differs in P0 and P1. P1 introduces the "enable_sync_mode" config to Wacc. `pack_config` should move to the Protocol_Px classes.

aedsinger replied:
I've removed the enable_sync stuff from Pimu and Wacc

hello-binit commented on body/stretch_body/wacc.py:
```diff
@@ -215,30 +179,110 @@ def pack_config(self,s,sidx):
             return sidx
 
+    def unpack_status(self,s):
+        pass
```
Would be better to `raise NotImplementedError()` here.

aedsinger replied:
Done

hello-binit commented on body/stretch_body/wacc.py:
```diff
@@ -44,43 +45,22 @@ def __init__(self, ext_status_cb=None, ext_command_cb=None):
         self.status = { 'ax':0,'ay':0,'az':0,'a0':0,'d0':0,'d1':0, 'd2':0,'d3':0,'single_tap_count': 0, 'state':0, 'debug':0,
                        'timestamp': 0,
                        'transport': self.transport.status}
+        self.ts_last=None
```
What does `ts_last` do?

aedsinger replied:
Vestigial variable from old timing stuff. Removed

hello-binit commented:
One more thought: would it make sense to, for a protocol X, include classes `Device_Protocol_P0`, `Device_Protocol_P1` to `Device_Protocol_PX`, as bases for `Device`? Suppose we add all waypoint traj functions to `Device_Protocol_P1`. If we only add `Device_Protocol_PX` as a base, then all the traj functions need to be copied forward.

Here's a MVP of what I'm thinking:

```python
In [3]: class PBase(): 
   ...:     def __init__(self): 
   ...:         pass 
   ...:     def say_hello_base(self): 
   ...:         print('hello base')

In [6]: class P0(): 
   ...:     def __init__(self): 
   ...:         pass 
   ...:     def say_hello_base(self): 
   ...:         print("hello base - p0") 
   ...:     def say_hello_p0(self): 
   ...:         print("hello p0 p0")

In [8]: class P1(): 
   ...:     def __init__(self): 
   ...:         pass 
   ...:     def say_hello_base(self): 
   ...:         print("hello base - p1") 
   ...:     def say_hello_p1(self): 
   ...:         print("hello p1 p1") 

In [26]: class P(PBase): 
    ...:     def __init__(self): 
    ...:         P.__bases__ = (P1, P0,) 

In [27]: o = P()                                                                                                                                                                                                                                                                          

In [28]: o.say_hello_base()                                                                                                                                                                                                                                                               
hello base - p1

In [29]: o.say_hello_p0()                                                                                                                                                                                                                                                                 
hello p0 p0

In [30]: o.say_hello_p1()                                                                                                                                                                                                                                                                 
hello p1 p1
```

hello-binit commented on body/stretch_body/stepper.py:
```diff
@@ -656,18 +608,18 @@ def pack_gains(self,s,sidx):
-                config=config | CONFIG_SAFETY_HOLD
+                config=config | self.CONFIG_SAFETY_HOLD
             if self.gains['enable_runstop']:
                 config=config | CONFIG_ENABLE_RUNSTOP
```
`CONFIG_ENABLE_RUNSTOP` should be `self.CONFIG_ENABLE_RUNSTOP` here.

hello-binit replied:
Nvm, fixed this and another constant that missed the "self".

aedsinger commented on body/stretch_body/pimu.py:
```diff
@@ -114,10 +81,45 @@ def unpack_status(self, s):
         return sidx
```
Ok, I've moved IMU to have base/P0/P1

aedsinger commented:
> One more thought: would it make sense to, for a protocol X, include classes `Device_Protocol_P0`, `Device_Protocol_P1` to `Device_Protocol_PX`, as bases for `Device`? Suppose we add all waypoint traj functions to `Device_Protocol_P1`. If we only add `Device_Protocol_PX` as a base, then all the traj functions need to be copied forward.
> 
> Here's a MVP of what I'm thinking:
> 
> ```python
> In [3]: class PBase(): 
>    ...:     def __init__(self): 
>    ...:         pass 
>    ...:     def say_hello_base(self): 
>    ...:         print('hello base')
>    ...:     def say_hello_p0(self): 
>    ...:         raise NotImplementedError()
>    ...:     def say_hello_p1(self): 
>    ...:         raise NotImplementedError()
> 
> In [6]: class P0(): 
>    ...:     def __init__(self): 
>    ...:         pass 
>    ...:     def say_hello_base(self): 
>    ...:         print("hello base - p0") 
>    ...:     def say_hello_p0(self): 
>    ...:         print("hello p0 p0")
> 
> In [8]: class P1(): 
>    ...:     def __init__(self): 
>    ...:         pass 
>    ...:     def say_hello_base(self): 
>    ...:         print("hello base - p1") 
>    ...:     def say_hello_p1(self): 
>    ...:         print("hello p1 p1") 
> 
> In [26]: class P(PBase): 
>     ...:     def __init__(self): 
>     ...:         P.__bases__ = (P1, P0,) 
> 
> In [27]: o = P()                                                                                                                                                                                                                                                                          
> 
> In [28]: o.say_hello_base()                                                                                                                                                                                                                                                               
> hello base - p1
> 
> In [29]: o.say_hello_p0()                                                                                                                                                                                                                                                                 
> hello p0 p0
> 
> In [30]: o.say_hello_p1()                                                                                                                                                                                                                                                                 
> hello p1 p1
> ```

Its a good point - one that I had considered. Conversely, if we go with this approach then there is no way to deprecate a method that is available in P0 if you are running P1+, for example. This is where things get tricky, as it won't be clear what the actual API for the device is (it will have to be worked out by construction).

 I suppose in P1 we could just implement the deprecated method with a ``` raise DeprecatedInterfaceError()```

hello-binit commented on body/stretch_body/pimu.py:
```diff
@@ -488,3 +473,114 @@ def step_sentry(self):
+        if self.hw_valid:
+            if self.board_info['protocol_version'] in self.supported_protocols:
+                Pimu.__bases__ = (self.supported_protocols[self.board_info['protocol_version']],)
+                self.imu.__bases__= (self.imu.supported_protocols[self.board_info['protocol_version']],)
```
Should be `IMU.__bases__` instead of `self.imu.__bases__`. Fails with NotImplementedError on the latter.

aedsinger replied:
Fixed

aedsinger commented:
Ok, I've updated so that the protocols override eachother as suggested. I've also cleaned up and removed the leftover enable_sync_mode code from both the firmware and Stretch Body.

hello-binit reviewed (approved):
Looks good!
